asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py

```diff
@@ -0,0 +1,378 @@
+        description="If the net charge of the chemical system should be neutralized by the ions defined by `positive_ion` abd `negative_ion`.",
+    )
+    ion_concentration: FloatQuantity["molar"] = Field(  # noqa: F821
+        0.15 * OFFUnit.molar,
+        description="The ionic concentration required in molar units.",
+    )
+
+    def to_solvent_component(self) -> gufe.SolventComponent:
+        return gufe.SolventComponent(**self.dict(exclude={"type"}))
+
+
+# TODO make base class with abstract methods to collect results.
+class TransformationResult(_SchemaBaseFrozen):
+    """
+    Store the results of a transformation, note when retries are used this will be the average result.
+    """
+
+    type: Literal["TransformationResult"] = "TransformationResult"
+    ligand_a: str = Field(
+        ..., description="The name of the ligand in state A of the transformation."
+    )
+    ligand_b: str = Field(
+        ..., description="The name of the ligand in state B of the transformation."
+    )
+    phase: Literal["complex", "solvent"] = Field(
+        ..., description="The phase of the transformation."
+    )
+    estimate: FloatQuantity["kcal/mol"] = Field(  # noqa: F821
+        ..., description="The average estimate of this transformation in kcal/mol"
+    )
+    uncertainty: FloatQuantity["kcal/mol"] = Field(  # noqa: F821
+        ...,
+        description="The standard deviation of the estimates of this transform in kcal/mol",
+    )
+
+    def name(self):
+        """Make a name for this transformation based on the names of the ligands."""
+        return "-".join([self.ligand_a, self.ligand_b])
+
+
+class _BaseResults(_SchemaBaseFrozen):
+    """
+    A base results class which handles the collecting and processing of the results.
+    """
+
+    type: Literal["_BaseResults"] = "_BaseResults"
+    results: list[TransformationResult] = Field(
+        [], description="The list of results collected for this dataset."
+    )
+
+    def to_cinnabar_measurements(self):
+        """
+        For the given set of results combine the solvent and complex phases to make a list of cinnabar RelativeMeasurements
+
+        Returns:
+            A list of RelativeMeasurements made from the combined solvent and complex phases.
+        """
+        from collections import defaultdict
+
+        import numpy as np
+        from cinnabar import RelativeMeasurement
+
+        raw_results = defaultdict(list)
+        # gather by transform
+        for result in self.results:
+            raw_results[result.name()].append(result)
+
+        # make sure we have a solvent and complex phase for each result
+        for name, transforms in raw_results.items():
+            missing_phase = {"complex", "solvent"} - {t.phase for t in transforms}
+            if missing_phase:
+                raise RuntimeError(
+                    f"The transformation {name} is missing simulated legs in the following phases {missing_phase}"
+                )
```

Comment:
As your require a 2 length iterable in   ` for leg1, leg2 in raw_results.values():` below, could also check that it is a 2 length iterable. I know this is a bit pedantic.

Reply:
Good catch added a check for this as well and a separate test.